realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -1048,6 +1081,151 @@ void BaseRealSenseNode::publishStaticTransforms()
     }
 }
 
+/* Modified for Brain */
+static void output_unit_vector(unsigned int height, unsigned int width, float unitVectors[FRAME_MAX_HEIGHT][FRAME_MAX_WIDTH * 3])
+{
+	unsigned int i, j;
+	FILE *fp; /* output to .obj file with a certain format to be shown by some tool */
+
+	std::cerr << "\nUnit Test for Realsense Unit Vector:\nOutput the saved data: frame width is " \
+		<< width << " and frame height is " << height << " \n\n" << std::endl;
+
+	std::cerr << "*** Beginning of unit vectors ***\n" << std::endl;
+
+	fp = fopen("./realsense_unit_vectors.obj","w");
+	if (!fp) {
+		std::cerr << "\nFailed to open a file for realsense_unit_vectors.obj!\n\n" << std::endl;
+		return;
+	}
+
+	std::cerr << "\nOutput unit_vectors to ./realsense_unit_vectors.obj!\n\n" << std::endl;
+
+	for (i = 0; i < height; i++)
+		for (j = 0; j < width * 3; j = j + 3)
+			fprintf(fp, "v %.6f %.6f %.6f\n", unitVectors[i][j], unitVectors[i][j + 1], unitVectors[i][j + 2]);
+
+	fclose(fp);
+
+	std::cerr << "\n*** End of unit vectors ***\n\n" << std::endl;
+}
+
+void BaseRealSenseNode::publishUnitVectorsTopic(const ros::Time& t, const std::map<stream_index_pair, bool>& is_frame_arrived)
+{
+	static float num_valid;
+	static bool unit_vector_mature;
+	static unsigned int height = 240, width = 424, sequence;
+	static float current_unit_vectors[FRAME_PRIME_HEIGHT][FRAME_PRIME_WIDTH * 3]= {0};
+
+	try
+	{
+		if (!is_frame_arrived.at(DEPTH))
+		{
+			ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't arrive.");
+			return;
+		}
+	}
+	catch (std::out_of_range)
+	{
+		ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't configure.");
+		return;
+	}
+
+	auto image_depth16 = reinterpret_cast<const uint16_t*>(_image[DEPTH].data);
```

Comment:
So _image is a CV Mat with an underlying data type of uint8 to make it non-specific to any stream type. How we access that data depends on the stream type (i.e. we need to reinterpret the data uint16 for depth images).

The combination of a raw pointer and _reinterpret_cast_ seems... a little dangerous. Rather than using a pointer to the data, consider using [CV Mat's builtin templated _at_ or _begin_](https://docs.opencv.org/3.4/d3/d63/classcv_1_1Mat.html) functions with uint16 as the template argument.